WordPress/gutenberg
# Testing: Reject release on unremoved deprecations

Related: #5069, #5398

This pull request seeks to fail the build via ESLint error when deprecations marked for removal in a given version change are not removed. The intention here (and as explained at https://github.com/WordPress/gutenberg/pull/5069#issuecomment-366717259) is that the success of a deprecation schedule depends on both clear communication and strict enforcement. A lack of predictability erodes trust and discourages adherence.

__Implementation notes:__

There appear to be more changes than there really are because the ESLint configuration needed to be ported from JSON to JavaScript. The changes introduced by this pull request are isolated to 3b2d1ce.

__Testing instructions:__

Verify that after manually changing the `package.json` version to `2.4.0`, running the lint task produces errors:

```
npm run lint
```

```
> gutenberg@2.4.0 lint /Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg
> eslint .


/Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/blocks/block-description/index.js
  15:4  error  Deprecated functions must be removed before releasing this version  no-restricted-syntax

/Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/blocks/hooks/matchers.js
  14:3  error  Deprecated functions must be removed before releasing this version  no-restricted-syntax

/Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/blocks/inspector-controls/index.js
  35:5  error  Deprecated functions must be removed before releasing this version  no-restricted-syntax

/Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/blocks/media-upload/button.js
  16:4  error  Deprecated functions must be removed before releasing this version  no-restricted-syntax

/Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/data/index.js
  146:4  error  Deprecated functions must be removed before releasing this version  no-restricted-syntax

✖ 5 problems (5 errors, 0 warnings)
```

__Follow-up Tasks:__

- With the ESLint configuration ported to JavaScript, it may be easier to update the E2E-specific lint rules to extend this as a base (see 8b1d6f12e0890dcd6b8e7798163518a8da0d0f9f, #4836)
   - Related to this, we should enforce ESLint rules within our the ESLint file itself. This requires excluding it from `.eslintignore` via `!.eslintrc.js`, but this currently produces errors in `test/e2e/.eslintrc.js`
